go-controller/pkg/ovn/libovsdbops/acl.go

```diff
@@ -0,0 +1,220 @@
+package libovsdbops
+
+import (
+	"fmt"
+	"reflect"
+
+	libovsdbclient "github.com/ovn-org/libovsdb/client"
+	libovsdb "github.com/ovn-org/libovsdb/ovsdb"
+
+	"github.com/ovn-org/ovn-kubernetes/go-controller/pkg/nbdb"
+)
+
+// getACLName returns the ACL name if it has one otherwise returns
+// an empty string.
+func getACLName(acl *nbdb.ACL) string {
+	if acl.Name != nil {
+		return *acl.Name
+	}
+	return ""
+}
+
+// isEquivalentACL if it hase same uuid, or if it has same name
+// and external ids, or if it has same priority, direction, match
+// and action.
+func isEquivalentACL(existing *nbdb.ACL, searched *nbdb.ACL) bool {
+	if searched.UUID != "" && existing.UUID == searched.UUID {
+		return true
+	}
+
+	eName := getACLName(existing)
+	sName := getACLName(searched)
+	// TODO if we want to support adding/removing external ids,
+	// we need to compare them differently, perhaps just the common subset
+	if eName != "" && eName == sName && reflect.DeepEqual(existing.ExternalIDs, searched.ExternalIDs) {
+		return true
+	}
+
+	return existing.Priority == searched.Priority &&
+		existing.Direction == searched.Direction &&
+		existing.Match == searched.Match &&
+		existing.Action == searched.Action
+}
+
+// findACL looks up the client cache for equivalent ACLs
+func findACL(nbClient libovsdbclient.Client, acl *nbdb.ACL) (string, error) {
+	if acl.UUID != "" && !isNamedUUID(acl.UUID) {
+		return acl.UUID, nil
+	}
+
+	acls := []nbdb.ACL{}
+	err := nbClient.WhereCache(func(item *nbdb.ACL) bool {
+		return isEquivalentACL(item, acl)
+	}).List(&acls)
+	if err != nil && err != libovsdbclient.ErrNotFound {
+		return "", fmt.Errorf("can't find ACL by equivalence %+v: %v", *acl, err)
+	}
+
+	if len(acls) > 1 {
+		return "", fmt.Errorf("unexpectedly found multiple equivalent ACLs: %+v", acls)
+	}
+
+	if len(acls) == 1 {
+		return acls[0].UUID, nil
+	}
+
+	return "", nil
+}
+
+func BuildACL(name, direction string, priority int, match, action, meter, severity string, log bool, externalIds map[string]string) *nbdb.ACL {
+	uuid := buildNamedUUID()
+	name = fmt.Sprintf("%.63s", name)
+	return &nbdb.ACL{
+		UUID:        uuid,
+		Name:        &name,
+		Direction:   direction,
+		Match:       match,
+		Action:      action,
+		Priority:    priority,
+		Severity:    &severity,
+		Log:         log,
+		Meter:       &meter,
+		ExternalIDs: externalIds,
+	}
+}
+
+func CreateOrUpdateACLOps(nbClient libovsdbclient.Client, ops []libovsdb.Operation, acl *nbdb.ACL) ([]libovsdb.Operation, error) {
+	if ops == nil {
+		ops = []libovsdb.Operation{}
+	}
+
+	uuid, err := findACL(nbClient, acl)
+	if err != nil {
+		return nil, err
+	}
+
+	// If ACL already exists, update it
+	if uuid != "" {
+		acl.UUID = ""
```

Comment:
is setting the UUID empty correct?

Reply:
Yeah, for updates, if it is not empty, libovsdb will complain. Other choice is to specify which attributes we want to update, so we would need to list them all except uuid.